onnxruntime/python/dlpack_convertor.cc

```diff
@@ -0,0 +1,234 @@
+    case DLDeviceType::kDLGPU:
+    case DLDeviceType::kDLROCM:
+      return OrtDevice(OrtDevice::GPU, OrtDevice::MemType::DEFAULT, static_cast<OrtDevice::DeviceId>(ctx.device_id));
+    default:
+      ORT_THROW("Unsupported device type");
+  }
+}
+
+MLDataType GetOrtValueDataType(const DLDataType& dtype) {
+  if (dtype.lanes != 1) ORT_THROW("ORT does not support lanes != 1");
+  switch (dtype.code) {
+    case DLDataTypeCode::kDLUInt:
+      switch (dtype.bits) {
+        case 8:
+          return DataTypeImpl::GetType<uint8_t>();
+        case 16:
+          return DataTypeImpl::GetType<uint16_t>();
+        case 32:
+          return DataTypeImpl::GetType<uint32_t>();
+        case 64:
+          return DataTypeImpl::GetType<uint64_t>();
+        default:
+          ORT_THROW("Unsupported kUInt bits " + std::to_string(dtype.bits));
+      }
+    case DLDataTypeCode::kDLInt:
+      switch (dtype.bits) {
+        case 8:
+          return DataTypeImpl::GetType<int8_t>();
+        case 16:
+          return DataTypeImpl::GetType<int16_t>();
+        case 32:
+          return DataTypeImpl::GetType<int32_t>();
+        case 64:
+          return DataTypeImpl::GetType<int64_t>();
+        default:
+          ORT_THROW("Unsupported kInt bits " + std::to_string(dtype.bits));
+      }
+    case DLDataTypeCode::kDLFloat:
+      switch (dtype.bits) {
+        case 16:
+          return DataTypeImpl::GetType<MLFloat16>();
+        case 32:
+          return DataTypeImpl::GetType<float>();
+        case 64:
+          return DataTypeImpl::GetType<double>();
+        default:
+          ORT_THROW("Unsupported kFloat bits " + std::to_string(dtype.bits));
+      }
+    default:
+      ORT_THROW("Unsupported code " + std::to_string(dtype.code));
+  }
+}
+
+const char* GetOrtDeviceName(const OrtDevice& device) {
+  switch (device.Type()) {
+    case OrtDevice::CPU:
+      return CPU;
+    case OrtDevice::GPU:
+      return CUDA;
+    default:
+      ORT_THROW("Unknown device type: ", device.Type());
+  }
+}
+
+bool IsContiguousTensor(const DLTensor& tensor) {
+  if (!tensor.strides) {
+    return true;
+  }
+
+  int64_t running_size = 1;
+  for (int i = tensor.ndim - 1; i >= 0; i--) {
+    if (tensor.strides[i] != running_size) {
+      return false;
+    }
+
+    running_size *= tensor.shape[i];
+  }
+
+  return true;
+}
+
+}  // namespace
+
+// This function returns a shared_ptr to memory managed DLpack tensor
+// constructed out of OrtValue.
+DLManagedTensor* OrtValueToDlpack(const OrtValue& ort_value) {
+  ORT_ENFORCE(ort_value.IsTensor(), "Only tensor type OrtValues are supported");
+  OrtDLManagedTensor* ort_dlmanaged_tensor(new OrtDLManagedTensor);
+  const Tensor& tensor = ort_value.Get<Tensor>();
+  ort_dlmanaged_tensor->handle = ort_value;
+  ort_dlmanaged_tensor->tensor.manager_ctx = ort_dlmanaged_tensor;
+  ort_dlmanaged_tensor->tensor.deleter = &DlpackDeleter;
+  ort_dlmanaged_tensor->tensor.dl_tensor.data = const_cast<void*>(tensor.DataRaw());
```

Comment:
>const_cast [](start = 48, length = 10)

is this const_cast necessary? will the returned DLManagedTensor* be used to modify the data? if so, should ort_value not be const?

Reply:
ort_value can not be const here. 
updated.